storage/merge.go

```diff
@@ -333,10 +333,23 @@ func (c *genericMergeSeriesSet) Next() bool {
 	for {
 		// Firstly advance all the current series sets. If any of them have run out,
 		// we can drop them, otherwise they should be inserted back into the heap.
-		for _, set := range c.currentSets {
-			if set.Next() {
-				heap.Push(&c.heap, set)
-			}
+		ch := make(chan genericSeriesSet, len(c.currentSets))
+		wg := sync.WaitGroup{}
+		for i := range c.currentSets {
+			wg.Add(1)
+			i := i
+			go func() {
+				defer wg.Done()
+				if c.currentSets[i].Next() {
+					ch <- c.currentSets[i]
+				}
+			}()
+		}
+
+		wg.Wait()
+		close(ch)
+		for set := range ch {
+			heap.Push(&c.heap, set)
```

Comment:
Can we make this change in a separate pr? And I think it might not be required to change it upstream if we really need this capability in Thanos/Cortex.
Can we just specify a different merge func here https://github.com/prometheus/prometheus/blob/main/tsdb/compact.go#L84?